### DISCUSSION TITLE
Rust: "No definition found" for method of cloned variable

### DISCUSSION THREAD

**Discussion by VasilisManol** [helix-editor/helix#11244] [Troubleshooting] [answered]
### Summary

I was looking through helix source code, specifically helix-term/src/commands.rs:2337  from master branch (c9d829a26dac81a57f18796308d3806d77d0c439) and noticed that line nr.2337 `"let mut searcher = searcher.clone();"`
 has no inlay hint. Further examining it looks like it doesn't get any type from the LSP. If going then to line nr.2371 `"if searcher.multi_line_with_matcher(&matcher) {" ` then go to definition doesn't work and gives "no definition found". I thought that it might be a short coming of rust-analyzer but the same file in vscode with rust-analyzer extension works as expected.
![screenshot-2024-07-19-18-57-56](https://github.com/user-attachments/assets/e1be9af2-644d-4030-bf11-63bab0957f11)

### Reproduction Steps

- Install hx 24.7 (from release page, I have tested with the previous version as well and the same error exists) with default config
- Checkout https://github.com/helix-editor/helix master branch
- Install rust toolchain and rust-analyzer with default instructions for linux
- Go to file helix-term/src/commands.rs:2337  and turn inlay hints on with :toggle-option lsp.display-inlay-hints
- Notice that no inlay hint is shown.
- Go to first reference of this variable (line 2371) and try go to definition

### Helix log

[helix.log](https://github.com/user-attachments/files/16316462/helix.log)

### Platform

Linux

### Terminal Emulator

alacritty 0.12.2

### Installation Method

binary from release page

### Helix Version

helix 24.7 (079f5442)

**Comment by kirawi** [comment, marked as answer]
I'm not able to replicate this with rust-analyzer 2024-07-14. Your rust-analyzer version is very old, 2023-11-13. You should try upgrading.

**Reply by VasilisManol** [reply]
Thanks! Using a newer rust-analyzer solves the problem. However, rust-analyzer 2023-11-13 is the one that comes by default with rustup and toolchain 1.74.0 that is specified in the rust-toolchain.toml file in helix repo. I think best compatibility is with the version in the toolchain as written here https://rust-analyzer.github.io/manual.html#toolchain. Nevertheless,  rust-analyzer from stable toolchain works fine so far. For reference, to force rust-analyzer to a another version you need to specify the RUSTUP_TOOLCHAIN env variable. In  helix languages.toml add:

```toml
[language-server.rust-analyzer]
environment = { "RUSTUP_TOOLCHAIN" = "stable"}
```
